python/ray/data/datatype.py

```diff
@@ -0,0 +1,217 @@
+from dataclasses import dataclass
+from typing import Any, Union
+
+import numpy as np
+import pyarrow as pa
+
+from ray.util.annotations import PublicAPI
+
+
+def generate_arrow_factory_methods(cls: type):
+    """Metaprogrammming: Class decorator to generate factory methods for PyArrow types using from_arrow."""
+
+    # Define the mapping of method name -> (pyarrow_func, description)
+    type_definitions = {
+        "int8": (pa.int8, "an 8-bit signed integer"),
+        "int16": (pa.int16, "a 16-bit signed integer"),
+        "int32": (pa.int32, "a 32-bit signed integer"),
+        "int64": (pa.int64, "a 64-bit signed integer"),
+        "uint8": (pa.uint8, "an 8-bit unsigned integer"),
+        "uint16": (pa.uint16, "a 16-bit unsigned integer"),
+        "uint32": (pa.uint32, "a 32-bit unsigned integer"),
+        "uint64": (pa.uint64, "a 64-bit unsigned integer"),
+        "float32": (pa.float32, "a 32-bit floating point number"),
+        "float64": (pa.float64, "a 64-bit floating point number"),
+        "string": (pa.string, "a variable-length string"),
+        "bool": (pa.bool_, "a boolean value"),
+        "binary": (pa.binary, "variable-length binary data"),
+    }
+
+    for method_name, (pa_func, description) in type_definitions.items():
+
+        def create_method(name, func, desc):
+            def factory_method(cls):
+                return cls.from_arrow(func())
+
+            factory_method.__doc__ = f"""Create a DataType representing {desc}.
+
+        Returns:
+            DataType: A DataType with PyArrow {name} type
+        """
+            factory_method.__name__ = name
+            factory_method.__qualname__ = f"{cls.__name__}.{name}"
+            return classmethod(factory_method)
+
+        setattr(cls, method_name, create_method(method_name, pa_func, description))
+
+    return cls
+
+
+@PublicAPI(stability="alpha")
+@dataclass
+@generate_arrow_factory_methods
+class DataType:
+    """A simplified Ray Data DataType supporting Arrow, NumPy, and Python types."""
+
+    _internal_type: Union[pa.DataType, np.dtype, type]
+
+    def __post_init__(self):
+        """Validate the _internal_type after initialization."""
+        if not isinstance(self._internal_type, (pa.DataType, np.dtype, type)):
```

Comment:
We need to match support of batch UDFs: ie we'd have to support PA, Numpy, Pandas, Python (not a fan of the latter but since we already support it)